examples/stream/src/main.rs

```diff
@@ -0,0 +1,140 @@
+            EnvFilter::builder()
+                .with_default_directive(LevelFilter::INFO.into())
+                .from_env()?,
+        )
+        .init();
+
+    let maybe_address = std::env::args()
+        .nth(1)
+        .map(|arg| arg.parse::<Multiaddr>())
+        .transpose()
+        .context("Failed to parse argument as `Multiaddr`")?;
+
+    let mut swarm = libp2p::SwarmBuilder::with_new_identity()
+        .with_tokio()
+        .with_quic()
+        .with_behaviour(|_| stream::Behaviour::default())?
+        .with_swarm_config(|c| c.with_idle_connection_timeout(Duration::from_secs(10)))
+        .build();
+
+    swarm.listen_on("/ip4/127.0.0.1/udp/0/quic-v1".parse()?)?;
+
+    // Register our custom protocol with the behaviour.
+    let (control, mut incoming_streams) = swarm.behaviour_mut().register(PROTOCOL).unwrap();
+
+    // Deal with incoming streams.
+    // Spawning a dedicated task is just one way of doing this.
+    // libp2p doesn't care how you handle incoming streams but you _must_ handle them somehow.
+    // To mitigate DoS attacks, libp2p will internally drop incoming streams if your application cannot keep in processing them.
+    tokio::spawn(async move {
+        // This loop handles incoming streams _sequentially_ but that doesn't have to be the case.
+        // You can also spawn a dedicated task per stream if you want to.
+        // Be aware that this breaks backpressure though as spawning new tasks is equivalent to an unbounded buffer.
+        // Each task needs memory meaning an aggressive remote peer may force you OOM this way.
+
+        while let Some((peer, stream)) = incoming_streams.next().await {
+            if let Err(e) = inbound_ping(stream).await {
+                tracing::warn!(%peer, "Ping protocol failed: {e}");
+                continue;
+            }
+
+            tracing::info!(%peer, "Handled inbound ping!");
+        }
+    });
+
+    // In this demo application, the dialing peer initiates the protocol.
+    if let Some(address) = maybe_address {
+        let Some(Protocol::P2p(peer_id)) = address.iter().last() else {
+            anyhow::bail!("Provided address does not end in `/p2p`");
+        };
+
+        swarm.dial(address)?;
+
+        tokio::spawn(connection_handler(peer_id, control));
+    }
+
+    // Poll the swarm to make progress.
+    loop {
+        let event = swarm.next().await.expect("never terminates");
+
+        match event {
+            libp2p::swarm::SwarmEvent::NewListenAddr { address, .. } => {
+                let listen_address = address.with_p2p(*swarm.local_peer_id()).unwrap();
+                tracing::info!(%listen_address);
+            }
+            event => tracing::trace!(?event),
+        }
+    }
+}
+
+/// A very simple, `async fn`-based connection handler for our custom ping protocol.
+async fn connection_handler(peer: PeerId, mut control: stream::Control) {
+    let mut peer_control = control.peer(peer).await.unwrap();
```

Comment:
Why do we need the additional indirection of a peer control? Why not allow a user to create a stream off of the `stream::Control` directly?

Reply:
Without the indirection, garbage collection is also easier, right? E.g. what if all `Control`s of a protocol are dropped but a `PeerControl` for the user protocol still exists?

Reply:
That is what I had at first but I think it makes upholding backpressure complicated: The message contains the `PeerId` so I need to read from the channel before I know which channel to a connection handler would want the message but that channel might be full! So in order to uphold backpressure, I now need to buffer this message and stop polling the channel until I can send the other one. As a result, the slowest connection will hold back all others.

For this protocol, all the handler needs to do is open a stream so one might argue, it doesn't matter that much but I just don't find it a very clean solution.

Do you have any ideas on how else to solve this?

An unrelated aspect that also made me go for the `PeerControl` was that it could be nice extension point if one wants to add the ability to e.g. control the dialing by adding addresses to it. However, I opted not to implement that for simplicity and I am not convinced it is something that this behaviour should deal with.

Reply:
Something else I just came across is connection keep-alive. My current plan is to keep connections alive while `PeerControl`s exist. Do you think that is a good idea? On the one hand, it gives users _some_ control over managing connections. On the other hand, there is built-in functionality to just dial the remote peer when we want to send them a message.

Reply:
Doing the keep-alive is actually somewhat tricky and with the receiver's being shared across all connections, it isn't very clean because now _all_ connections would be kept alive.

I think I'll remove that again.

The point above still stands though. I don't know how else to model it cleanly other than with the indirection of `PeerControl`. Ideas welcome!